#### Loading the libraries

In [1]:
# Importing the required libraries

import pandas as pd
import numpy as np

#### Loading the dataset

In [2]:
# Loading the dataset

df = pd.read_csv("../data/raw/dynamic_pricing.csv")

#### Displaying the first 10 rows of the dataset to get a quick look at the data

In [3]:
df.head(10)

,Number_of_Riders,Number_of_Drivers,Location_Category,Customer_Loyalty_Status,Number_of_Past_Rides,Average_Ratings,Time_of_Booking,Vehicle_Type,Expected_Ride_Duration,Historical_Cost_of_Ride
0,90,45,Urban,Silver,13,4.47,Night,Premium,90,284.257273
1,58,39,Suburban,Silver,72,4.06,Evening,Economy,43,173.874753
2,42,31,Rural,Silver,0,3.99,Afternoon,Premium,76,329.795469
3,89,28,Rural,Regular,67,4.31,Afternoon,Premium,134,470.201232
4,78,22,Rural,Regular,74,3.77,Afternoon,Economy,149,579.681422
5,59,35,Urban,Silver,83,3.51,Night,Economy,128,339.955361
6,93,43,Suburban,Regular,44,4.41,Afternoon,Premium,16,104.061541
7,62,39,Rural,Gold,83,3.59,Afternoon,Premium,47,235.811864
8,79,14,Rural,Silver,71,3.74,Evening,Economy,128,501.412517
9,42,6,Rural,Silver,21,3.85,Night,Premium,128,398.993365


#### Displaying the total number of rows and columns in the dataset

In [4]:
rows, columns = df.shape
print(f'The dataset contains {rows} rows and {columns} columns.')

The dataset contains 1000 rows and 10 columns.


#### Displaying column names

In [5]:
df.columns

Index(['Number_of_Riders', 'Number_of_Drivers', 'Location_Category',
       'Customer_Loyalty_Status', 'Number_of_Past_Rides', 'Average_Ratings',
       'Time_of_Booking', 'Vehicle_Type', 'Expected_Ride_Duration',
       'Historical_Cost_of_Ride'],
      dtype='object')

In [6]:
# Checking the data types

df.dtypes

Number_of_Riders             int64
Number_of_Drivers            int64
Location_Category           object
Customer_Loyalty_Status     object
Number_of_Past_Rides         int64
Average_Ratings            float64
Time_of_Booking             object
Vehicle_Type                object
Expected_Ride_Duration       int64
Historical_Cost_of_Ride    float64
dtype: object

#### Based on the dataset documentation, the meaning of each column in the Dynamic Pricing dataset is as follows:

* Number of Riders - Number of riders requesting rides under the given market conditions.
* Number of Drivers - Number of drivers available to provide rides under the given market conditions.
* Location Category - Category representing the type of location associated with the ride, such as urban, suburban, or rural.
* Customer Loyalty Status - Loyalty category assigned to the customer based on their relationship with the ride-sharing service.
* Number of Past Rides - Number of rides previously taken by the customer.
* Average Ratings - Average rating associated with the customer or their previous ride activity.
* Time of Booking - Time period or category representing when the ride was booked.
* Vehicle Type - Type or category of vehicle selected for the ride.
* Expected Ride Duration - Expected duration of the ride.
* Historical Cost of Ride - Historical fare or cost associated with the ride, serving as the primary target variable for ride-cost prediction.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Number_of_Riders         1000 non-null   int64  
 1   Number_of_Drivers        1000 non-null   int64  
 2   Location_Category        1000 non-null   object 
 3   Customer_Loyalty_Status  1000 non-null   object 
 4   Number_of_Past_Rides     1000 non-null   int64  
 5   Average_Ratings          1000 non-null   float64
 6   Time_of_Booking          1000 non-null   object 
 7   Vehicle_Type             1000 non-null   object 
 8   Expected_Ride_Duration   1000 non-null   int64  
 9   Historical_Cost_of_Ride  1000 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 78.3+ KB


In [8]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Number_of_Riders,1000.0,60.372000,23.701506,20.000000,40.000000,60.000000,81.000000,100.000000
Number_of_Drivers,1000.0,27.076000,19.068346,5.000000,11.000000,22.000000,38.000000,89.000000
Number_of_Past_Rides,1000.0,50.031000,29.313774,0.000000,25.000000,51.000000,75.000000,100.000000
Average_Ratings,1000.0,4.257220,0.435781,3.500000,3.870000,4.270000,4.632500,5.000000
Expected_Ride_Duration,1000.0,99.588000,49.165450,10.000000,59.750000,102.000000,143.000000,180.000000
Historical_Cost_of_Ride,1000.0,372.502623,187.158756,25.993449,221.365202,362.019426,510.497504,836.116419


#### Displaying columns along with their data types

In [9]:
print(df.dtypes)

Number_of_Riders             int64
Number_of_Drivers            int64
Location_Category           object
Customer_Loyalty_Status     object
Number_of_Past_Rides         int64
Average_Ratings            float64
Time_of_Booking             object
Vehicle_Type                object
Expected_Ride_Duration       int64
Historical_Cost_of_Ride    float64
dtype: object


#### Obersvations on Data Types

The dataset contains numerical variables stored as int64 or float64 and categorical variables stored as object. The current data types are appropriate for their respective variables

In [10]:
for col in df.columns:
    unique_vals = df[col].unique()
    print(f'Column Name : {col}')
    print(f'Number of Unique Values : {len(unique_vals)}')
    print(f'Unique Values : {unique_vals}')
    print('-'*50)

Column Name : Number_of_Riders
Number of Unique Values : 81
Unique Values : [ 90  58  42  89  78  59  93  62  79  86  60  24  36  63  39  38  46  33
  31  96  22  74  48  56  72  97  57  25  75  35  68  28  73  29  94  54
  67  23  70  20  95  43  32  47  37  53  40  76  41  83  44  91  45  52
  81  77  27  92  66  61  71  49  55  34  51  88  26  80  64  84 100  21
  65  82  85  69  50  87  99  30  98]
--------------------------------------------------
Column Name : Number_of_Drivers
Number of Unique Values : 79
Unique Values : [45 39 31 28 22 35 43 14  6 17 38  8 24 13 12  7 11 16 19 33 75 56  9 32
 15 18 10 81 29 84 55 20 25 64 23 40 26 73 37 34 69 21 44 30 42  5 47 49
 51 48 36 46 80 59 41 50 57 71 62 27 63 67 60 54 53 66 52 72 76 65 58 74
 61 70 87 77 89 85 82]
--------------------------------------------------
Column Name : Location_Category
Number of Unique Values : 3
Unique Values : ['Urban' 'Suburban' 'Rural']
--------------------------------------------------
Column Name : Cus

#### Checking for missing values per column (including empty strings)

In [11]:
# Count of NaN / None values per column
nan_count = df.isnull().sum()

# Count of empty strings per column
empty_count = (df == '').sum()

# Total missing values (NaN + empty strings)
total_missing = nan_count + empty_count

# Percentage of missing values
missing_percentage = (total_missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': total_missing,
    'Percentage (%)': missing_percentage
}, index=df.columns).sort_values(by='Percentage (%)', ascending=False)

print(missing_df)

                         Missing Values  Percentage (%)
Number_of_Riders                      0             0.0
Number_of_Drivers                     0             0.0
Location_Category                     0             0.0
Customer_Loyalty_Status               0             0.0
Number_of_Past_Rides                  0             0.0
Average_Ratings                       0             0.0
Time_of_Booking                       0             0.0
Vehicle_Type                          0             0.0
Expected_Ride_Duration                0             0.0
Historical_Cost_of_Ride               0             0.0


#### Checking for the duplicated rows.


In [12]:
df.duplicated().sum()

0

#### Numerical Data Validation

In [13]:
numerical_columns = [
    'Number_of_Riders',
    'Number_of_Drivers',
    'Number_of_Past_Rides',
    'Average_Ratings',
    'Expected_Ride_Duration',
    'Historical_Cost_of_Ride'
]

In [14]:
df[numerical_columns].describe()

,Number_of_Riders,Number_of_Drivers,Number_of_Past_Rides,Average_Ratings,Expected_Ride_Duration,Historical_Cost_of_Ride
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000
mean,60.372000,27.076000,50.031000,4.257220,99.58800,372.502623
std,23.701506,19.068346,29.313774,0.435781,49.16545,187.158756
min,20.000000,5.000000,0.000000,3.500000,10.00000,25.993449
25%,40.000000,11.000000,25.000000,3.870000,59.75000,221.365202
50%,60.000000,22.000000,51.000000,4.270000,102.00000,362.019426
75%,81.000000,38.000000,75.000000,4.632500,143.00000,510.497504
max,100.000000,89.000000,100.000000,5.000000,180.00000,836.116419


In [15]:
(df[numerical_columns] < 0).sum()

Number_of_Riders           0
Number_of_Drivers          0
Number_of_Past_Rides       0
Average_Ratings            0
Expected_Ride_Duration     0
Historical_Cost_of_Ride    0
dtype: int64

#### Date / Time Validation

In [17]:
print("Unique Time_of_Booking values:")
print(df['Time_of_Booking'].unique())

Unique Time_of_Booking values:
['Night' 'Evening' 'Afternoon' 'Morning']


In [18]:
print("\nFrequency of Time_of_Booking:")
print(df['Time_of_Booking'].value_counts())


Frequency of Time_of_Booking:
Time_of_Booking
Night        276
Afternoon    247
Morning      246
Evening      231
Name: count, dtype: int64


#### categorical data validation

In [19]:
categorical_columns = [
    'Location_Category',
    'Customer_Loyalty_Status',
    'Time_of_Booking',
    'Vehicle_Type'
]

for col in categorical_columns:
    print(f'\n{col}:')
    print(df[col].value_counts())


Location_Category:
Location_Category
Urban       346
Rural       332
Suburban    322
Name: count, dtype: int64

Customer_Loyalty_Status:
Customer_Loyalty_Status
Silver     367
Regular    320
Gold       313
Name: count, dtype: int64

Time_of_Booking:
Time_of_Booking
Night        276
Afternoon    247
Morning      246
Evening      231
Name: count, dtype: int64

Vehicle_Type:
Vehicle_Type
Premium    522
Economy    478
Name: count, dtype: int64


#### Target Variable Validation

In [20]:
print("Target Variable: Historical_Cost_of_Ride")
print("\nTarget Data Type:", df['Historical_Cost_of_Ride'].dtype)
print("Missing Values:", df['Historical_Cost_of_Ride'].isnull().sum())
print("Unique Values:", df['Historical_Cost_of_Ride'].nunique())

Target Variable: Historical_Cost_of_Ride

Target Data Type: float64
Missing Values: 0
Unique Values: 1000


In [21]:
df['Historical_Cost_of_Ride'].describe()

count    1000.000000
mean      372.502623
std       187.158756
min        25.993449
25%       221.365202
50%       362.019426
75%       510.497504
max       836.116419
Name: Historical_Cost_of_Ride, dtype: float64

#### Observations on Numerical Data

All numerical variables contain valid non-negative values, with no negative observations detected. The variables also fall within their observed ranges, with `Average_Ratings` ranging from 3.5 to 5.0 and `Expected_Ride_Duration` ranging from 10 to 180 minutes.


#### Observations on Time of Booking

The `Time_of_Booking` variable contains four valid booking periods: `Morning`, `Afternoon`, `Evening`, and `Night`. All four categories are represented in the dataset, with no unexpected or missing time-period values.


#### Observations on Categorical Data

The categorical variables contain a limited number of well-defined categories, with all expected categories represented in the dataset. The category distributions are reasonably balanced, and no extremely rare categories requiring consolidation or removal were identified.


#### Observations on Target Variable

The target variable `Historical_Cost_of_Ride` is stored as a `float64` numerical variable with no missing values. All 1,000 observations have unique ride-cost values, indicating continuous variation in the target and confirming its suitability for regression analysis.